File: Java/24. Swap Nodes in Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

 // Iterative Solution
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode ans = new ListNode(0);
        ans.next = head;
        ListNode prev = ans;
        
        while(head != null && head.next != null){
            ListNode second = head.next;
            
            prev.next = second;
            head.next = second.next;
            second.next = head;
            
            prev = head;
            head = head.next;
        }
        
        return ans.next;
    }
}

File: Python3/24. Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Recursive Solution
class Solution:
    def swapPairs(self, first: Optional[ListNode]) -> Optional[ListNode]:
        
        if not first or not first.next:
            return first
        
        second = first.next
        
        # before: first -> second -> second.next
        # after:  second -> first -> second.next
        second.next, first.next = first, self.swapPairs(second.next)
        
        return second